File: examples/python/opaque_modes.py
#!/usr/bin/env python3
"""Tiny opaque Python workload demonstrating all three result modes."""

import argparse
from pathlib import Path

from structured_result import encode_envelope, output_entry, publish_atomic, succeeded_payload


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("mode", choices=("exit-only", "declared-output", "structured"))
    parser.add_argument("workspace", type=Path)
    args = parser.parse_args()

    if args.mode == "exit-only":
        return

    output = args.workspace / "results" / "answer.txt"
    output.parent.mkdir(parents=True, exist_ok=True)
    output.write_text("42\n", encoding="utf-8")
    if args.mode == "declared-output":
        return

    payload = succeeded_payload(
        submission_key="python-opaque-example",
        attempt_id="python-opaque-attempt",
        attempt_epoch=1,
        operation={"kind": "script", "sourceDigest": "sha256:" + "c" * 64},
        result_schema="example.answer.v1",
        value=b'{"answer":42}',
        outputs=[output_entry(args.workspace, "results/answer.txt")],
        worker_release={"id": "python-helper-1", "digest": "sha256:" + "d" * 64},
        completed_at="2026-07-22T12:00:00Z",
    )
    publish_atomic(args.workspace / "result.json", encode_envelope(payload))


if __name__ == "__main__":
    main()

File: examples/R/opaque_modes.R
# Tiny opaque R workload demonstrating exit-only, declared-output, and structured modes.
args <- commandArgs(trailingOnly = TRUE)
if (length(args) != 2L) stop("usage: Rscript opaque_modes.R MODE WORKSPACE")
mode <- args[[1L]]
workspace <- args[[2L]]
if (!mode %in% c("exit-only", "declared-output", "structured")) stop("unknown mode")
if (mode == "exit-only") quit(status = 0L)

output <- file.path(workspace, "results", "answer.txt")
dir.create(dirname(output), recursive = TRUE, showWarnings = FALSE)
writeLines("42", output, useBytes = TRUE)
if (mode == "declared-output") quit(status = 0L)

script_arg <- commandArgs(trailingOnly = FALSE)
script_file <- sub("^--file=", "", script_arg[grepl("^--file=", script_arg)][[1L]])
source(file.path(dirname(normalizePath(script_file)), "structured_result.R"))
payload <- succeeded_payload(
  "r-opaque-example",
  "r-opaque-attempt",
  1L,
  list(kind = "script", sourceDigest = paste0("sha256:", strrep("c", 64L))),
  "example.answer.v1",
  charToRaw('{"answer":42}'),
  list(output_entry(workspace, "results/answer.txt")),
  list(id = "r-helper-1", digest = paste0("sha256:", strrep("d", 64L))),
  "2026-07-22T12:00:00Z"
)
publish_atomic(file.path(workspace, "result.json"), encode_envelope(payload))
